service/create/asg_stack.go

```diff
@@ -246,16 +257,55 @@ func (s *Service) updateASGStack(input asgStackInput) error {
 		Client: input.clients.CloudFormation,
 
 		// Settings.
-		ASGMaxSize:  input.asgSize,
-		ASGMinSize:  input.asgSize,
-		ImageID:     imageID,
-		Name:        key.AutoScalingGroupName(input.cluster, input.asgType),
-		TemplateURL: templateURL,
+		ASGMaxSize:            input.asgSize,
+		ASGMinSize:            input.asgSize,
+		ImageID:               imageID,
+		MaxBatchSize:          getMaxBatchSize(input.asgSize),
+		MinInstancesInService: getMinInstancesInService(input.asgSize),
+		Name: key.AutoScalingGroupName(input.cluster, input.asgType),
+		RollingUpdatePauseTime: rollingUpdatePauseTime,
+		TemplateURL:            templateURL,
+	}
+
+	stackUpdated, err := stack.Update()
+	if err != nil {
+		return stackUpdated, microerror.Mask(err)
+	}
+
+	return stackUpdated, nil
+}
+
+// getMaxBatchSize calculates the max batch size for the rolling update policy.
+func getMaxBatchSize(asgSize int) int {
+	var batchSize int
+
+	switch {
+	case asgSize <= 2:
+		batchSize = 1
+	case asgSize <= 4:
+		batchSize = 2
+	default:
+		// Calculate batch size and round up to nearest int.
+		result := float64(asgSize) * asgMaxBatchSizeRatio
+		batchSize = int(result + 0.5)
 	}
 
-	if err := stack.Update(); err != nil {
-		return microerror.Mask(err)
+	return batchSize
+}
+
+// getMinInstancesInService calculates the min number for instances to keep in
+// service for the rolling update policy.
+func getMinInstancesInService(asgSize int) int {
+	var minInstances int
+
+	switch {
+	case asgSize <= 2:
+		minInstances = 1
+	default:
+		// Calculate min instances and round up to nearest int.
+		result := float64(asgSize) * asgMinInstancesRatio
+		minInstances = int(result + 0.5)
```

Comment:
Ok, this is big. I thought about how to implement rolling updates and updating CoreOS versions of guest clusters. I mainly focused on KVM. IMO the operator implements a reconciliation of the resources it obtains and only provides primitives to update guest cluster details. Such a detail can be adding or removing a node. As I see it right now the operation of rolling the complete guest cluster is a collection of operations that the operator itself should not manage itself. Thinking about KVM adding and removing a node means also changing the TPO/CRO. The TPO/CRO defines the desired state. Implementing the batch processing of a rolling update in the operator implies the desired state has to be changed by the operator itself, which feels super weird. I am not sure how this is intended to work but having an operator modifying its own TPO/CRO feels wrong. I am also wondering how this should work in AWS. Right now it looks like the references of the nodes being tracked in the TPO/CRO stay the same while the nodes are actually changing, which makes the use of node IDs pretty much useless. What is your thinking here? I am also interested in @kopiczko's opinion here. IMO the creation and synchronization of the batch of operations should be done within the authority that manages the TPO/CRO, which is `kubernetesd` in our case. So what I actually would like to do in KVM would be to add a rolling updates mechanism in kubernetesd which updates the necessary bits for the CoreOS version and then adds a node to the TPO/CRO, waits for it to come up, removes an old node and continues doing so until the whole guest cluster is rolled. Having the mechanisms of how to roll and update a cluster in the operator feels kind of wrong because it essentially means it screws around with its own TPO/CRO, which it should not be able to modify. I understand that kubernetesd might be the wrong place for the batch processing though, especially because it is closed source and the community cannot benefit from it. What we can do though is providing an upgrade controller that does all the magic. This can then be open source and be nicely integrated into all kinds of environments while keeping the operator clean. What I want to prevent is modifying the TPO/CRO within a reconciliation framework resource implementation. A resource implementation must only be concerned with managing itself.